File: packages/program-design/alog/数据结构/二叉树/py/二叉树展开为链表.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        if not root:
            return root
        self.flatten(root.left)
        self.flatten(root.right)
        # 后序遍历，能够保证自底向上 & 编辑每一个子树节点
        left = root.left
        right = root.right

        root.left = None
        root.right = left
        
        p = root
        while p.right:
            p = p.right
        p.right = right

File: packages/program-design/alog/数据结构/二叉树/ts/二叉树展开为链表.ts
import {TreeNode} from './types'

// 前序遍历
function flatten(root: TreeNode | null): void {
    if (root === null) {
        return;
    }

    let current = new TreeNode(0); // 定义一个辅助节点
    //前序遍历的原因是题目要求前序遍历
    const preTraverse = (node: TreeNode | null) => {
        if (node === null) {
            return;
        }

        const left = node.left;
        const right = node.right;

        current.right = node;
        current.left = null;
        current = node;

        preTraverse(left);
        preTraverse(right);
    }

    preTraverse(root);
};